Implement dotnet nuget sign command

heng-liu opened the pull request:
<!-- DO NOT MODIFY OR DELETE THIS TEMPLATE. IT IS USED IN AUTOMATION. -->
## Bug

<!-- Search https://github.com/NuGet/Home/issues, and create one if you can't find a suitable issue. -->
<!-- Paste the full link, like https://github.com/nuget/home/issues/1000. GitHub will render is neatly. -->
Fixes: https://github.com/NuGet/Home/issues/10581

Regression? Last working version:

## Description
1. Implement dotnet nuget sign command (This is done by @kartheekp-ms)
2. Add tests in test/NuGet.Clients.Tests/NuGet.CommandLine.Test/NuGetSignCommandTest.cs 
3. Add tests in NuGet.XPlat.FuncTest/XplatSignTests.cs 
4. Add tests in Dotnet.Integration.Test/DotnetSignTests.cs (This is partly done by @kartheekp-ms)
5. Change the signing test fixture a little bit, to set up a 3 level certs to sign, as on Mac, a self-issued trusted cert needs to use many commands to add to trusted store, according to  https://github.com/dotnet/runtime/issues/49528#issuecomment-801159398
6. Fix the issue of NU3001 error code not showing in dotnet.exe, by handle the exception earlier. Fix tests accordingly.

Note:
After offline discussion with @aortiz-msft and @JonDouglas, the
--interactive and other formats of certificate (like .pem) will not be supported in this PR.

Those two will be implemented in the future insertion. The tracking issues are:
https://github.com/NuGet/Home/issues/10620
https://github.com/NuGet/Home/issues/10626

## PR Checklist

- [x] PR has a meaningful title
- [x] PR has a linked issue.
- [x] Described changes

- **Tests**
  - [x] Automated tests added
  - **OR**
  <!-- Describe why you haven't added automation. -->
  - [ ] Test exception
  - **OR**
  - [ ] N/A <!-- Infrastructure, documentation etc. -->

- **Documentation**
  <!-- Please link the PR/issue if appropriate -->
  - [x] Documentation PR or issue filled  https://github.com/NuGet/docs.microsoft.com-nuget/issues/2416
  - **OR**
  - [ ] N/A

kartheekp-ms committed:
initial sign

kartheekp-ms committed:
start dotnet sign tests

heng-liu committed:
add some unit test and fix a bug

heng-liu committed:
add tests into SignCommandRunnerTests

AraHaan commented:
I would like the pfx to be used on non-windows systems if the codebase only supplies a pfx file for signing.

erdembayar commented on src/NuGet.Core/NuGet.CommandLine.XPlat/Commands/Signing/SignCommand.cs:
```diff
@@ -0,0 +1,239 @@
+                    "--interactive",
+                    Strings.Verbosity_Description,
+                    CommandOptionType.NoValue);
+
```
Where is `--configfile`? I can see it on spec.

heng-liu replied:
Good question.
I checked there is no config file in [SignCommand ](https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Clients/NuGet.CommandLine/Commands/SignCommand.cs), [SignArgs ](https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Core/NuGet.Commands/SignCommand/SignArgs.cs) or [SignCommandRunner](https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Core/NuGet.Commands/SignCommand/SignCommandRunner.cs). So NuGet.exe doesn't have -configFile either. 
I also went through all configurations in https://docs.microsoft.com/en-us/nuget/reference/nuget-config-file, none of them will be used during the sign command. 
So I guess sign command doesn't need config file at all.. We might need to remove that option from the spec.
Hi @kartheekp-ms , seems dotnet verify command doesn't have config file either, right? Do you remember the reason? Thanks!

heng-liu replied:
Thanks @kartheekp-ms for offline confirmation.
So the following document need to be fixed:
https://docs.microsoft.com/en-us/nuget/reference/cli-reference/cli-ref-sign
https://docs.microsoft.com/en-us/nuget/reference/cli-reference/cli-ref-verify
The dotnet command spec needs to remove -configfile option as well.

erdembayar replied:
Could you be able to create tracking issue for fixing document/spec and paste link here?

heng-liu replied:
Thanks~ Raised doc issue : https://github.com/NuGet/docs.microsoft.com-nuget/issues/2416

erdembayar commented on src/NuGet.Core/NuGet.CommandLine.XPlat/Commands/Signing/SignCommand.cs:
```diff
@@ -0,0 +1,239 @@
+            {
+                string outputDir = output.Value();
+
+                if (!!Directory.Exists(outputDir))
```
This double negation doesn't look doing anything useful. 
Did you mean?
`if (!Directory.Exists(outputDir))`

heng-liu replied:
You're right. Fixed. Thanks.

erdembayar commented on src/NuGet.Core/NuGet.CommandLine.XPlat/Commands/Signing/SignCommand.cs:
```diff
@@ -0,0 +1,239 @@
+        {
+            if (output.HasValue())
+            {
+                string outputDir = output.Value();
```
Do you have relative path as argument unit test? Mostly cli people user assumes paths as relative as current location.

heng-liu replied:
Good point! 
Added tests `DotnetSign_SignPackageWithPfxFileOfRelativePath_SuccessAsync` and `DotnetSign_SignPackageWithTrustedCertificateWithRelativePath_SuccceedsAsync`.

erdembayar commented on src/NuGet.Core/NuGet.CommandLine.XPlat/Commands/Signing/SignCommand.cs:
```diff
@@ -0,0 +1,239 @@
+                CommandArgument packagePaths = signCmd.Argument(
+                    "<package-paths>",
+                    Strings.SignCommandPackagePathDescription,
+                    multipleValues: true);
```
Do you have relative path as argument unit test? Mostly cli people user assumes paths as relative as current location.

heng-liu replied:
Same as above.
Added tests DotnetSign_SignPackageWithPfxFileOfRelativePath_SuccessAsync and DotnetSign_SignPackageWithTrustedCertificateWithRelativePath_SuccceedsAsync.

erdembayar commented on src/NuGet.Core/NuGet.CommandLine.XPlat/Commands/Signing/SignCommand.cs:
```diff
@@ -0,0 +1,239 @@
+                throw new ArgumentException(Strings.SignCommandNoCertificateException);
+            }
+            else if (!string.IsNullOrEmpty(path.Value()) &&
+                ((!string.IsNullOrEmpty(fingerPrint.Value()) ||
```
Do we really need this double `(` ?
Not sure if it's contributing any value? Maybe fix it for nuget.exe too.

heng-liu replied:
Thanks! Fixed that in both nuget.exe and dotnet.exe.

erdembayar commented on src/NuGet.Core/NuGet.CommandLine.XPlat/Utility/CommandLineUtility.cs:
```diff
@@ -36,5 +39,31 @@ public static string[] SplitAndJoinAcrossMultipleValues(IList<string> inputs)
+        /// <param name="optionName">Name of the command line argument</param>
+        /// <param name="spec">Signing specification to validate parsed hash algorithm</param>
+        /// <returns>Supported hash algorithm</returns>
+        internal static HashAlgorithmName ParseAndValidateHashAlgorithm(string optionValue, string optionName, SigningSpecifications spec)
```
Why only this function is here while others does very similar things still in `SignCommand.cs`?
As I can see this one only used in `SignCommand.cs`.

heng-liu replied:
My understanding is it's following the same pattern with https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Clients/NuGet.CommandLine/Common/CommandLineUtility.cs#L54
ParseAndValidateHashAlgorithmFromArgument is parsing and validating HashAlgorithm. It looks like a general method that other command might use, but currently only sign command uses it right now.

I could move that method from a utility to signcommand class for both nuget.exe and dotnet.exe, if it's needed.

erdembayar commented on src/NuGet.Core/NuGet.Commands/SignCommand/SignArgs.cs:
```diff
@@ -16,8 +18,33 @@ public class SignArgs
+        {
+            get
+            {
+                switch (PackagePaths.Count)
```
Do we need null check here?

kartheekp-ms replied:
`PackagePaths` is an argument. It is mandatory.

heng-liu replied:
Thanks @kartheekp-ms , so there is no need for the null check here, right?

kartheekp-ms replied:
Yes no null check is required, because validations were performed before populating `SignArgs`.

heng-liu replied:
Thanks @kartheekp-ms !

erdembayar commented on src/NuGet.Core/NuGet.Commands/SignCommand/SignCommandRunner.cs:
```diff
@@ -22,8 +23,12 @@ public class SignCommandRunner : ISignCommandRunner
-            LocalFolderUtility.EnsurePackageFileExists(signArgs.PackagePath, packagesToSign);
+            var packagesToSign = signArgs.PackagePaths.SelectMany(packagePath =>
+            {
+                var packages = LocalFolderUtility.ResolvePackageFromPath(packagePath);
```
Explicit type please. It looks `IEnumerable<string>` returning here.

heng-liu replied:
Thanks. Fixed.

erdembayar commented on test/NuGet.Core.FuncTests/Dotnet.Integration.Test/DotnetSignTests.cs:
```diff
@@ -0,0 +1,497 @@
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
```
Remove this `using System.Collections.Generic;`, not used.

erdembayar replied:
Also below too:
using System.IO.Compression;
using System.Linq;
using System.Text;

heng-liu replied:
Thanks! Fixed.

erdembayar commented on test/NuGet.Core.FuncTests/Dotnet.Integration.Test/DotnetSignTests.cs:
```diff
@@ -0,0 +1,497 @@
+using System.Linq;
+using System.Security.Cryptography;
+using System.Security.Cryptography.X509Certificates;
+#if IS_SIGNING_SUPPORTED
```
Do we need this since it looks this cs file only target `netcoreapp5.0`?
`#if IS_SIGNING_SUPPORTED`

heng-liu replied:
You're right. Removed.

erdembayar commented on test/NuGet.Core.FuncTests/Dotnet.Integration.Test/DotnetSignTests.cs:
```diff
@@ -0,0 +1,497 @@
+
+                var packageFilePath = Path.Combine(pathContext.PackageSource, "PackageA.1.0.0.nupkg");
+
+                var trustedCert = _signFixture.TrustedTestCertificateChain.Leaf;
```
Explicit type please?

heng-liu replied:
Thanks! Fixed.

erdembayar commented on test/NuGet.Core.FuncTests/Dotnet.Integration.Test/SignCommandTestFixture.cs:
```diff
@@ -0,0 +1,381 @@
+            {
+                if (_trustedTestCertChain == null)
+                {
+                    var certChain = SigningTestUtility.GenerateCertificateChain(_validCertChainLength, CrlServer.Uri, TestDirectory.Path);
```
Explicit type here please. Same for other cert chains.

heng-liu replied:
Thanks! Fixed.

erdembayar commented on test/NuGet.Core.FuncTests/Dotnet.Integration.Test/SignCommandTestFixture.cs:
```diff
@@ -0,0 +1,381 @@
+                if (_trustedTestCertWithInvalidEku == null)
+                {
+                    var actionGenerator = SigningTestUtility.CertificateModificationGeneratorForInvalidEkuCert;
+                    var certChain = SigningTestUtility.GenerateCertificateChain(_validCertChainLength, CrlServer.Uri, TestDirectory.Path, configureLeafCrl: true, leafCertificateActionGenerator: actionGenerator);
```
Looks many duplicate codes between those methods. Can we reduce it?

heng-liu replied:
I checked the pattern and add comments as following. There seems not many duplicate code between each type of certificate. 
```
                if (_trustedTestCertWithInvalidEku == null)  // check if the private property( a specific type of certficate) is null
                {
                    var actionGenerator = SigningTestUtility.CertificateModificationGeneratorForInvalidEkuCert;  // get the specific action generator
                    IList<TrustedTestCert<TestCertificate>> certChain = SigningTestUtility.GenerateCertificateChain(_normalCertChainLength, CrlServer.Uri, TestDirectory.Path, configureLeafCrl: true, leafCertificateActionGenerator: actionGenerator);  //generate the list of certificates in the chain by passing some parameters

                    TrustedTestCertificateChain _trustedTestLeafCertWithInvalidEkuChain = new TrustedTestCertificateChain()
                    {
                        Certificates = certChain
                    };  // new a TrustedTestCertificateChain by passing the list

                    SetUpCrlDistributionPoint();

                    _trustedTestCertWithInvalidEku = _trustedTestLeafCertWithInvalidEkuChain.Leaf;  //assign the leaf certificate of the chain to the property
                }
```

erdembayar commented on test/NuGet.Core.FuncTests/Dotnet.Integration.Test/DotnetSignTests.cs:
```diff
@@ -0,0 +1,497 @@
+
+                var trustedLeafCert = _signFixture.TrustedTestCertificateChain.Leaf;
+                //Act
+                var result = _msbuildFixture.RunDotnet(
```
This one looks duplicate of line#39 test. Please correct me if I'm wrong.

heng-liu replied:
Good find! Yes, they are the same. I removed this one. Thanks!

erdembayar commented on test/NuGet.Core.FuncTests/Dotnet.Integration.Test/DotnetSignTests.cs:
```diff
@@ -0,0 +1,497 @@
+        }
+
+        [Fact]
+        public async Task DotnetSign_SignPackageWithTimestamping_SuccceedsAsync()
```
Triple `c` in `SuccceedsAsync`. Also same for all other `Succeed` tests.

heng-liu replied:
Thanks! Fixed.

erdembayar commented on test/NuGet.Core.FuncTests/Dotnet.Integration.Test/DotnetSignTests.cs:
```diff
@@ -0,0 +1,497 @@
+
+                // Assert
+                result.Success.Should().BeFalse(because: result.AllOutput);
+                //result.AllOutput.Should().Contain(string.Format(_invalidPasswordErrorCode, pfxPath));
```
Do we keep or remove the commented code?

kartheekp-ms replied:
I think we should remove it unless I am missing something.

heng-liu replied:
Thanks! Fixed the missing NU3001 error code issue, so this is fixed.

erdembayar commented on test/NuGet.Core.FuncTests/Dotnet.Integration.Test/DotnetSignTests.cs:
```diff
@@ -0,0 +1,497 @@
+                    ignoreExitCode: true);
+
+                // Assert
+                result.Success.Should().BeTrue(because: result.AllOutput);
```
Add `NU3018` warning too since it's self signed cert.

heng-liu replied:
Good catch! Fixed. Thanks.
BTW, it's untrusted self-issued cert. If we add the self-issued cert into trusted store, then it's trusted.

erdembayar commented on test/NuGet.Core.FuncTests/Dotnet.Integration.Test/DotnetSignTests.cs:
```diff
@@ -0,0 +1,497 @@
+                var originalFile = File.ReadAllBytes(packageFilePath);
+
+                var testServer = await _signFixture.GetSigningTestServerAsync();
+                var certificateAuthority = await _signFixture.GetDefaultTrustedCertificateAuthorityAsync();
```
Explicit type please.

heng-liu replied:
Thanks. Fixed.

erdembayar commented on test/NuGet.Core.FuncTests/Dotnet.Integration.Test/DotnetSignTests.cs:
```diff
@@ -0,0 +1,497 @@
+                //Act
+                var result = _msbuildFixture.RunDotnet(
+                    pathContext.PackageSource,
+                    $"nuget sign {packageFilePath} --certificate-fingerprint {untrustedSelfIssuedCert.Thumbprint}",
```
When we sign package with only using `Thumbprint` what is `Thumbprint` here?
Where is actual certificate coming from this case?

heng-liu replied:
Thumbprint is the same with fingerprint.
According to https://docs.microsoft.com/en-us/nuget/reference/cli-reference/cli-ref-sign, if we specify `CertificateFingerprint`, it specifies the SHA-1 fingerprint of the certificate used to search a local certificate store for the certificate.

erdembayar commented on test/NuGet.Core.FuncTests/Dotnet.Integration.Test/DotnetSignTests.cs:
```diff
@@ -0,0 +1,497 @@
+                    //Act
+                    var result = _msbuildFixture.RunDotnet(
+                        pathContext.PackageSource,
+                        $"nuget sign {packageFilePath} --certificate-fingerprint {UntrustedSelfIssuedCert.Thumbprint} --timestamper {timestampService.Url}",
```
What is difference between passing `Thumbprint` with and without `StoreName`/`StoreLocation`?
If we can pass just `Thumbprint` then what is the need for `StoreName`/`StoreLocation`

heng-liu replied:
If CertificatePath is specified, the cert will be from a certificate file.
If CertificateFingerprint/CertificateSubjectName/CertificateStoreLocation/CertificateStoreName is specified, it will search for a local certificate store.
If both of the two categories are specified, we will throw an exception.

If it's from local certificate store,
the default StoreName is StoreName.My.
the default StoreLocation is StoreLocation.CurrentUser.
If StoreName/StoreLocation is not specified, they will use the default value to search.
If the cert is not in My\CurrentUser, you have to specify the StoreName and StoreLocation.

erdembayar commented on test/NuGet.Core.FuncTests/Dotnet.Integration.Test/MockServer.cs:
```diff
@@ -0,0 +1,439 @@
+        /// <summary>
+        /// Initializes an instance of MockServer.
+        /// </summary>
+        public MockServer()
```
Can this one used for other integration/unit tests trying to make call to real nuget.org test?

erdembayar commented on test/NuGet.Core.FuncTests/Dotnet.Integration.Test/PortReserver.cs:
```diff
@@ -0,0 +1,205 @@
+    /// Gotcha: If another application grabs a port during the test, we have a race condition.
+    /// </summary>
+    [DebuggerDisplay("Port: {PortNumber}, Port count for this app domain: {_appDomainOwnedPorts.Count}")]
+    public class PortReserver : IDisposable
```
It looks this one is duplicate from `NuGet.CommandLine.Test`.
Not sure if it's possible, but can we have only 1 copy this instead of 2.

heng-liu replied:
Thanks for the question!
Actually, there are already two PortReservers in our repo. Both of them are used for testing. The first one has a more general usage and the second one is only used in MockServer(NuGet.CommandLine.Test)
https://github.com/NuGet/NuGet.Client/blob/dev/test/TestUtilities/Test.Utility/TestServer/PortReserver.cs
https://github.com/NuGet/NuGet.Client/blob/dev/test/NuGet.Clients.Tests/NuGet.CommandLine.Test/PortReserver.cs
There are following differences between the two:
1.APIs are different, the second one has BaseUri property
2.Ways of implementation: The first one tries to new a TcpListenner on that port to check if the port is occupied and uses ConcurrentDictionary to keep the reserved port if it's not occupied. The second one uses Mutex

And I need a third one for signing in Dotnet.integration.Test. I could not use the first one as there are different APIs. I could not use the second one as it's in NuGet.CommandLine.Test which only targets net472, while I need one targets net5.0.

I'll move the second one into Test.Utility as PortReserverOfMockServer.cs, so that both NuGet.CommandLine.Test and Dotnet.integration.Test could use this one. 

erdembayar commented on test/NuGet.Core.FuncTests/Dotnet.Integration.Test/SignCommandTestFixture.cs:
```diff
@@ -0,0 +1,381 @@
+            {
+                if (_revokedTestCertChain == null)
+                {
+                    var certChain = SigningTestUtility.GenerateCertificateChain(_invalidCertChainLength, CrlServer.Uri, TestDirectory.Path);
```
Does this one require `_invalidCertChainLength`? I believe it's only good for `RevocationUnknownTestCertificateWithChain`.

heng-liu replied:
Sorry for the confusion. I think `_invalidCertChainLength` might not be an accurate name. It should be `_shortCertChainLength`.
It doesn't imply the cert chain length is not valid, but setting up a short cert chain then it's easier to make it invalid.

erdembayar commented on test/NuGet.Core.FuncTests/Dotnet.Integration.Test/SignCommandTestFixture.cs:
```diff
@@ -0,0 +1,381 @@
+            {
+                if (_untrustedSelfIssuedCertificateInCertificateStore == null)
+                {
+                    var certificate = SigningTestUtility.GenerateSelfIssuedCertificate(isCa: false);
```
Just curious. Do we ever sign package with root CA certificate?
How about using intermediate certificate?

heng-liu replied:
Sorry for the late reply. You may already got the answer. We could use self-issued certificate to sign packages for testing. But those signed packages could not be uploaded to nuget.org. Nuget.org has validation pipeline to reject packages are signed by those certificate.
Pls refer to https://docs.microsoft.com/en-us/nuget/create-packages/sign-a-package#get-a-code-signing-certificate

erdembayar commented on test/NuGet.Core.FuncTests/Dotnet.Integration.Test/SignCommandTestFixture.cs:
```diff
@@ -0,0 +1,381 @@
+            _responders = new DisposableList<IDisposable>();
+        }
+
+        private void SetUpCrlDistributionPoint()
```
What `Crl` stand for? Is it abbreviation?

heng-liu replied:
Crl stands for certificate revocation list.
There are two ways to check the certificate revocation status: CRL and OCSP(Online Certificate Status Protocol).

erdembayar commented on test/NuGet.Core.Tests/NuGet.Commands.Test/SignCommandRunnerTests.cs:
```diff
@@ -81,28 +82,133 @@ public async Task ExecuteCommandAsync_WithNoCertificateFound_ThrowsAsync()
-        public async Task ExecuteCommandAsync_WithIncorrectPassword_ThrowsAsync()
+        public async Task ExecuteCommandAsync_WithExistingCertificateFromPathAndNoPassword_Succeed()
+        {
+            using (var test = await Test.CreateAsync(_fixture.GetDefaultCertificate()))
```
Explicit type please. Other it was hard to realize package is already created here.

heng-liu replied:
Thanks. I changed `Test `into `TestContext`.

erdembayar commented on test/TestUtilities/Test.Utility/Signing/CertificateStoreUtilities.cs:
```diff
@@ -18,5 +18,38 @@ public static StoreLocation GetTrustedCertificateStoreLocation(bool readOnly = f
+
+        internal static StoreName GetCertificateAuthorityStoreName()
+        {
+            // According to https://github.com/dotnet/runtime/issues/48207#issuecomment-778293907,
```
Above `dotnet comment` is about MAC. How about Linux? Does all of them exist in Linux?
        AddressBook = 1,
        AuthRoot = 2,
        CertificateAuthority = 3,
        Disallowed = 4,
        My = 5,
        Root = 6,
        TrustedPeople = 7,
        TrustedPublisher = 8

heng-liu replied:
Pls refer to https://github.com/dotnet/runtime/blob/main/docs/design/features/cross-platform-cryptography.md#x509store
`On Windows the X509Store class is a representation of the Windows Certificate Store APIs, and work the same as they did on .NET Framework. On Linux the X509Store class is a projection of system trust decisions (read-only), user trust decisions (read-write), and user key storage (read-write). On macOS the X509Store class is a projection of system trust decisions (read-only), user trust decisions (read-only), and user key storage (read-write).`

erdembayar commented on test/TestUtilities/Test.Utility/Signing/CertificateStoreUtilities.cs:
```diff
@@ -18,5 +18,38 @@ public static StoreLocation GetTrustedCertificateStoreLocation(bool readOnly = f
+            // only My, Root (RO), and Disallowed stores work on macOS.
+            // So use different approaches for Windows, Mac and Linux.
+
+            return (RuntimeEnvironmentHelper.IsWindows) ? StoreLocation.LocalMachine : StoreLocation.CurrentUser;
```
According above comment both `CurrentUser` and `LocalMachine` exist for MAC.
How about `Linux`?

heng-liu replied:
Good question!
Pls refer to https://github.com/dotnet/runtime/blob/main/docs/design/features/cross-platform-cryptography.md#x509store, it introduces the discrepancies between different platforms.
In our tests, in order to make the root certificate trusted, we need to add the root certificate to trusted store(so need to open trusted store with Write authority).
On Windows and Linux, we could do that by calling .NET APIs, but on MacOS,  the API could not be used as they will throw CryptographicException. So we run a command to add to trusted store on MacOS.

erdembayar commented on test/TestUtilities/Test.Utility/Signing/CertificateStoreUtilities.cs:
```diff
@@ -18,5 +18,38 @@ public static StoreLocation GetTrustedCertificateStoreLocation(bool readOnly = f
+            return (RuntimeEnvironmentHelper.IsWindows || RuntimeEnvironmentHelper.IsLinux) ? StoreName.CertificateAuthority : StoreName.My;
+        }
+
+        internal static StoreLocation GetCertificateAuthorityStoreLocation()
```
Looks like this method is unused. Are you going to keep it for future ref?

heng-liu replied:
Good catch! Removed.

erdembayar reviewed (commented):
Looks tremendous hard work. 👏 
I have left some comments. Please explicit actual type instead of `var`.

heng-liu committed:
address comments

erdembayar commented on src/NuGet.Core/NuGet.CommandLine.XPlat/Utility/CommandLineUtility.cs:
```diff
@@ -3,9 +3,12 @@
 using System.Collections.Generic;
+using System.Globalization;
+using System.Linq;
 using System.Text;
```
Please remove below import since not used: 
using System.Text;
using NuGet.Commands;

heng-liu replied:
Thanks! Fixed.

kartheekp-ms commented on src/NuGet.Core/NuGet.CommandLine.XPlat/Commands/Signing/SignCommand.cs:
```diff
@@ -0,0 +1,239 @@
+                        SignatureHashAlgorithm = hashAlgorithm,
+                        Logger = logger,
+                        Overwrite = overwrite.HasValue(),
+                        NonInteractive = !interactive.HasValue(),
```
Given that `interactive` option is low priority for now, should we remove it from the options list?

heng-liu replied:
You're right! Addressed. Thanks!

AraHaan commented:
Any plans to resolve the conflicts?

heng-liu commented:
> Any plans to resolve the conflicts?

Yes. I'm addressing the comments. Thanks for following up.

heng-liu committed:
address comment2

heng-liu committed:
fix

heng-liu committed:
address comments3

heng-liu commented:
Hi @erdembayar and @kartheekp-ms , thanks for both of you reviewing this PR.
I've addressed most of the comments. Pls let me know if there is any other comments. Thanks!

heng-liu committed:
fix missing NU3001 in dotnet.exe by handle exception early

kartheekp-ms commented on src/NuGet.Core/NuGet.CommandLine.XPlat/Commands/Signing/SignCommand.cs:
```diff
@@ -0,0 +1,241 @@
+
+        private static void ValidatePackagePaths(CommandArgument argument)
+        {
+            if (argument.Values.Count == 0)
```
how about adding an additional check to ensure empty strings are not passed as argument values?

https://github.com/NuGet/NuGet.Client/blob/3093ff5306ffddf9213fc61d89102e4afc8d9101/src/NuGet.Core/NuGet.CommandLine.XPlat/Commands/Signing/VerifyCommand.cs#L69-L72

heng-liu replied:
Thanks! Addressed.

kartheekp-ms commented on src/NuGet.Core/NuGet.CommandLine.XPlat/Strings.resx:
```diff
@@ -770,4 +770,59 @@ For more information, visit https://docs.nuget.org/docs/reference/command-line-r
+This option should be used when specifying the certificate via --certificate-subject-name or --certificate-fingerprint options.</value>
+  </data>
+  <data name="SignCommandCertificateStoreNameDescription" xml:space="preserve">
+    <value>Name of the X.509 certificate store to use to search for the certificate. Defaults to "My", the X.509 certificate store for personal certificates.
```
nit: We should add some extra information to `help` message because `the command accepts only one of the values defined in` https://docs.microsoft.com/en-us/dotnet/api/system.security.cryptography.x509certificates.storename `for StoreName option. `

kartheekp-ms replied:
nit: This comment applies to `SignCommandCertificateStoreNameDescription` string resource.

heng-liu replied:
According to https://github.com/dotnet/runtime/blob/main/docs/design/features/cross-platform-cryptography.md#x509store, the available StoreName and StoreLocation varies on different platforms.
It's so complicated that I'm afraid the information will be too much for the description? Or shall I link this section in the description?
BTW, `My` should be the common case if specifying certificate store name.

heng-liu replied:
How about adding the link of dotnet nuget sign document when the document is published?
Created tracking issue at NuGet/Home#10858

kartheekp-ms commented on src/NuGet.Core/NuGet.CommandLine.XPlat/Commands/Signing/SignCommand.cs:
```diff
@@ -0,0 +1,241 @@
+                    Strings.Verbosity_Description,
+                    CommandOptionType.SingleValue);
+
+                //The interactive option is not enabled at first. And it's tracked by https://github.com/NuGet/Home/issues/10620
+                //CommandOption interactive = signCmd.Option(
+                //    "--interactive",
+                //    Strings.Verbosity_Description,
+                //    CommandOptionType.NoValue);
```
nit: remove commented code because we have a tracking issue.

heng-liu replied:
Thanks! Removed~

kartheekp-ms commented on src/NuGet.Core/NuGet.CommandLine.XPlat/Commands/Signing/SignCommand.cs:
```diff
@@ -0,0 +1,241 @@
+                        TimestampHashAlgorithm = timestampHashAlgorithm
+                    };
+
+                    if (verbosity.HasValue())
```
If the default verbosity of `sign` command is `minimal` then you don't need this `if` statement.

https://github.com/NuGet/NuGet.Client/blob/3093ff5306ffddf9213fc61d89102e4afc8d9101/src/NuGet.Core/NuGet.CommandLine.XPlat/Commands/Signing/VerifyCommand.cs#L58

heng-liu replied:
Thanks! Addressed~

kartheekp-ms reviewed (commented):
I am still reviewing this PR but left few comments to start with.

heng-liu committed:
fix tests

kartheekp-ms commented on src/NuGet.Core/NuGet.CommandLine.XPlat/Strings.resx:
```diff
@@ -770,4 +770,59 @@ For more information, visit https://docs.nuget.org/docs/reference/command-line-r
+  </data>
+  <data name="SignCommandCertificatePasswordDescription" xml:space="preserve">
+    <value>Password for the certificate, if needed.
+This option can be used to specify the password for the certificate. If no password is provided, the command will prompt for a password at run time in the --interactive mode</value>
```
``` suggestion
This option can be used to specify the password for the certificate. The command will throw an error message if certificate is password protected but password is not provided as input.
```

heng-liu replied:
Good catch! Thanks @kartheekp-ms ! Fixed.

kartheekp-ms commented on src/NuGet.Core/NuGet.CommandLine.XPlat/Strings.resx:
```diff
@@ -770,4 +770,59 @@ For more information, visit https://docs.nuget.org/docs/reference/command-line-r
+    <value>Invalid value provided for '{0}'. The accepted values are {1}.</value>
+  </data>
+  <data name="SignCommandMultipleCertificateException" xml:space="preserve">
+    <value>Multiple options were used to specify a certificate.</value>
```
nit: Modify error message so that it is actionable. We should explain scenarios that can trigger this message https://github.com/NuGet/NuGet.Client/pull/3974/files#diff-4cf20fb6a1c739714683d0b642b396837f22ccf7152372552d134b73cfddcad3R215

heng-liu replied:
Considering the message for NuGet.exe is:
 [Multiple options were used to specify a certificate. For a list of accepted ways to provide a certificate, please visit https://docs.nuget.org/docs/reference/command-line-reference](https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Clients/NuGet.CommandLine/NuGetCommand.resx#L5481)

How about adding the link of dotnet nuget sign document when the document is published?
Created tracking issue at https://github.com/NuGet/Home/issues/10858

dominoFire commented on src/NuGet.Core/NuGet.CommandLine.XPlat/Strings.resx:
```diff
@@ -770,4 +770,59 @@ For more information, visit https://docs.nuget.org/docs/reference/command-line-r
+    <value>The '-Timestamper' option was not provided. The signed package will not be timestamped. To learn more about this option, please visit https://docs.microsoft.com/en-us/dotnet/core/tools</value>
+  </data>
+  <data name="Err_InvalidValue" xml:space="preserve">
+    <value>Invalid value provided for '{0}'. The accepted values are {1}.</value>
```
Please add a comment describing placeholders.

heng-liu replied:
Thanks. Added.

dominoFire commented on src/NuGet.Core/NuGet.CommandLine.XPlat/Strings.resx:
```diff
@@ -770,4 +770,59 @@ For more information, visit https://docs.nuget.org/docs/reference/command-line-r
+    <value>Signs NuGet package(s) at &lt;package-path&gt; with the specified certificate.</value>
+  </data>
+  <data name="SignCommandNoTimestamperWarning" xml:space="preserve">
+    <value>The '-Timestamper' option was not provided. The signed package will not be timestamped. To learn more about this option, please visit https://docs.microsoft.com/en-us/dotnet/core/tools</value>
```
Documentation URLs don't need to have locale. That locale is resolved on client machine.

```suggestion
    <value>The '-Timestamper' option was not provided. The signed package will not be timestamped. To learn more about this option, please visit https://docs.microsoft.com/dotnet/core/tools</value>
```

heng-liu replied:
Good to know! Thanks @dominoFire ~ Fixed.

dominoFire commented on src/NuGet.Core/NuGet.CommandLine.XPlat/Strings.resx:
```diff
@@ -770,4 +770,59 @@ For more information, visit https://docs.nuget.org/docs/reference/command-line-r
+  </data>
+  <data name="SignCommandCertificateSubjectNameDescription" xml:space="preserve">
+    <value>Subject name of the certificate used to search a local certificate store for the certificate.
+The search is a case-insensitive string comparison using the supplied value, which will find all certificates with the subject name containing that string, regardless of other subject values. The certificate store can be specified by --certificate-store-name and --certificate-store-location options.</value>
```
Please add a comment indication that `--certificate-store-name` and `--certificate-store-location` must not be translated. This also applies for all string resources that mention command options and/or flags.

heng-liu replied:
Sorry, @dominoFire , I didn't get it.. May I know why and how shall I add the comment? Thanks!

dominoFire replied:
Replying in a new comment..

dominoFire reviewed (commented):
Partial review. Added comments on string resources.

kartheekp-ms commented on test/NuGet.Core.FuncTests/Dotnet.Integration.Test/SignCommandTestFixture.cs:
```diff
@@ -0,0 +1,381 @@
+    /// <summary>
+    /// Used to bootstrap functional tests for signing.
+    /// </summary>
+    public class SignCommandTestFixture : IDisposable
```
I remember copying code from `NuGet.CommandLine.FuncTest/Commands/SignCommandTestFixture.cs` when I was working on the `dotnet sign` command. Can we move this class file to `TestUtility` so that the implementation can be shared in `NuGet.CommandLine.FuncTest` & `Dotnet.Integration.Test` projects without creating a new class?

https://github.com/NuGet/NuGet.Client/blob/dev/test/NuGet.Clients.FuncTests/NuGet.CommandLine.FuncTest/Commands/SignCommandTestFixture.cs

heng-liu replied:
Thanks @kartheekp-ms. Yes, at the beginning, the two fixtures are the same. 
But later on, I made some change in the test/NuGet.Core.FuncTests/Dotnet.Integration.Test/SignCommandTestFixture.cs, including:
1).Change the signing test fixture a little bit, to set up a 3 level certs to sign, as on Mac, a self-issued trusted cert needs to use many commands to add to trusted store, according to dotnet/runtime#49528 (comment)
2). Remove NuGetExePath as the dotnet integration tests doesn't need nuget.exe.
So the two are different now.
And I also found the test fixture usually stays with the test class in our repo. So I guess it's okey to not merge the two fixtures into test.utility?

dtivel replied:
If we can avoid duplication, that would be ideal.

1.  Can other tests adapt to the new 3-level certificate chains?
2.  NuGetExePath could just be ignored for dotnet tests.

Does this (from [here](https://xunit.net/docs/shared-context#collection-fixture)) require you to define a fixture in this assembly?
> Important note: Fixtures can be shared across assemblies, but collection definitions must be in the same assembly as the test that uses them.

heng-liu replied:
Thanks for pointing out that fixture could be shared between assemblies. Created a tracking issue https://github.com/NuGet/Client.Engineering/issues/950 to consolidate the two `SignCommandTestFixture.cs`.

heng-liu committed:
address comments 4

dominoFire commented on src/NuGet.Core/NuGet.CommandLine.XPlat/Strings.resx:
```diff
@@ -789,4 +789,61 @@ For more information, visit https://docs.nuget.org/docs/reference/command-line-r
+This option should be used when specifying the certificate via --certificate-subject-name or --certificate-fingerprint options.</value>
+  </data>
+  <data name="SignCommandCertificateSubjectNameDescription" xml:space="preserve">
+    <value>Subject name of the certificate used to search a local certificate store for the certificate.
+The search is a case-insensitive string comparison using the supplied value, which will find all certificates with the subject name containing that string, regardless of other subject values. The certificate store can be specified by --certificate-store-name and --certificate-store-location options.</value>
```
If an user tries to use the translated command option, the command will not work.

The comment is added in free text in the `<comment>` section

```suggestion
    <value>Subject name of the certificate used to search a local certificate store for the certificate.
The search is a case-insensitive string comparison using the supplied value, which will find all certificates with the subject name containing that string, regardless of other subject values. The certificate store can be specified by --certificate-store-name and --certificate-store-location options.</value>
    <comment>Please don't localize '--certificate-store-name' and '--certificate-store-location'</comment>
```

heng-liu replied:
I see.. Thanks! Good to know more about the localization work. Fixed.

dominoFire commented on test/TestUtilities/Test.Utility/Signing/TrustedTestCert.cs:
```diff
@@ -66,15 +66,20 @@ public class TrustedTestCert<T> : IDisposable
-                StoreLocation = storeLocation;
-                AddCertificateToStore();
+                AddCertificateToStoreForMacOSX();
+
```
nit: extra line

heng-liu replied:
Thanks~Fixed.

dominoFire commented on src/NuGet.Core/NuGet.CommandLine.XPlat/Strings.resx:
```diff
@@ -789,4 +789,61 @@ For more information, visit https://docs.nuget.org/docs/reference/command-line-r
+    <value>Switch to indicate if the current signature should be overwritten. By default the command will fail if the package already has a signature.</value>
+  </data>
+  <data name="SignCommandPackagePathDescription" xml:space="preserve">
+    <value>Signs NuGet packages at &lt;package-paths&gt; with the specified certificate.</value>
```
Add comment to not localize hard-coded strings indicating program arguments or keywords that need to be typed exactly as is.

```suggestion
    <value>Signs NuGet packages at &lt;package-paths&gt; with the specified certificate.</value>
    <comment>Please don't localize &lt;package-paths&gt;</comment>
```

heng-liu replied:
Thanks! Fixed.

dominoFire commented on src/NuGet.Core/NuGet.CommandLine.XPlat/Strings.resx:
```diff
@@ -789,4 +789,61 @@ For more information, visit https://docs.nuget.org/docs/reference/command-line-r
+    <value>Hash algorithm to be used by the RFC 3161 timestamp server. Defaults to SHA256.</value>
+  </data>
+  <data name="SignCommandDescription" xml:space="preserve">
+    <value>Signs NuGet package(s) at &lt;package-path&gt; with the specified certificate.</value>
```
```suggestion
    <value>Signs NuGet package(s) at &lt;package-path&gt; with the specified certificate.</value>
    <comment>Please don't localize '&lt;package-path&gt;'</comment>
```

heng-liu replied:
Thanks! Fixed.

dominoFire commented on src/NuGet.Core/NuGet.CommandLine.XPlat/Strings.resx:
```diff
@@ -789,4 +789,61 @@ For more information, visit https://docs.nuget.org/docs/reference/command-line-r
+    <value>Signs NuGet package(s) at &lt;package-path&gt; with the specified certificate.</value>
+  </data>
+  <data name="SignCommandNoTimestamperWarning" xml:space="preserve">
+    <value>The '-Timestamper' option was not provided. The signed package will not be timestamped. To learn more about this option, please visit https://docs.microsoft.com/dotnet/core/tools</value>
```
```suggestion
    <value>The '-Timestamper' option was not provided. The signed package will not be timestamped. To learn more about this option, please visit https://docs.microsoft.com/dotnet/core/tools</value>
    <comment>Please don't localize '-Timestamper'</comment>
```

heng-liu replied:
Thanks! Fixed.

dominoFire commented on src/NuGet.Core/NuGet.CommandLine.XPlat/Strings.resx:
```diff
@@ -789,4 +789,61 @@ For more information, visit https://docs.nuget.org/docs/reference/command-line-r
+    <value>Directory where the signed package(s) should be saved. By default the original package is overwritten by the signed package.</value>
+  </data>
+  <data name="SignCommandCertificateFingerprintDescription" xml:space="preserve">
+    <value>SHA-1 fingerprint of the certificate used to search a local certificate store for the certificate. The certificate store can be specified by --certificate-store-name and --certificate-store-location options.</value>
```
```suggestion
    <value>SHA-1 fingerprint of the certificate used to search a local certificate store for the certificate. The certificate store can be specified by --certificate-store-name and --certificate-store-location options.</value>
    <comment>Please don't localize '--certificate-store-name' and '--certificate-store-location'</comment>
```

heng-liu replied:
Thanks! Fixed.

dominoFire commented on src/NuGet.Core/NuGet.CommandLine.XPlat/Strings.resx:
```diff
@@ -789,4 +789,61 @@ For more information, visit https://docs.nuget.org/docs/reference/command-line-r
+    <value>File path to the certificate to be used while signing the package.</value>
+  </data>
+  <data name="SignCommandCertificateStoreLocationDescription" xml:space="preserve">
+    <value>Name of the X.509 certificate store use to search for the certificate. Defaults to "CurrentUser", the X.509 certificate store used by the current user.
+This option should be used when specifying the certificate via --certificate-subject-name or --certificate-fingerprint options.</value>
```
```suggestion
    <value>Name of the X.509 certificate store use to search for the certificate. Defaults to "CurrentUser", the X.509 certificate store used by the current user.
This option should be used when specifying the certificate via --certificate-subject-name or --certificate-fingerprint options.</value>
    <comment>Please don't localize '--certificate-subject-name' and '--certificate-fingerprint'</comment>
```

heng-liu replied:
Thanks! Fixed.

dominoFire commented on test/NuGet.Core.FuncTests/Dotnet.Integration.Test/MockServer.cs:
```diff
@@ -0,0 +1,440 @@
+    /// A Mock Server that is used to mimic a NuGet Server.
+    /// </summary>
+    [System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Design", "CA1001:TypesThatOwnDisposableFieldsShouldBeDisposable")]
+    public class MockServer : IDisposable
```
This class looks very similar to NuGet.CommandLine.Test [`MockServer`](https://github.com/NuGet/NuGet.Client/blob/73dac0ecc5172895c25d20b152e636504fb9751d/test/NuGet.Clients.Tests/NuGet.CommandLine.Test/MockServer.cs#L25), minus three OData methods: `ToODataFeed()`, `ToODataEntryXElement()`, `ToOData()`.

Can this code be shared in `Test.Utility` or in another specialized command line test utility project? @zivkan could have some ideas on `Test.Utility` usage.

heng-liu replied:
I tried to address this comment, while the effort is relatively bigger than I expect. Since it only affects the testing, can I do it gradually,  in a separate PR?

I tried in branch dev-hengliu-xplat3-feature-dotnet-nuget-sign-test2,  to consolidate the two MockServer, into Test.Utilities project.
The MockServer under NuGet.CommandLine.Test needs an extra NuGet.Core package  https://github.com/NuGet/NuGet.Client/blob/dev/test/NuGet.Clients.Tests/NuGet.CommandLine.Test/NuGet.CommandLine.Test.csproj#L12-L15, while when I tried to add that package to Test.Utilities.csproj, there are many conflicts, e.g. ILogger is previously from NuGet.Common, but now it's from the newly added NuGet.Core package.
I haven't looked into why we need the old NuGet.Core package. But will do that in a separate PR if possible.

Pls let me know if it's okey. If yes, I'll create a tracking issue to refactor the code to consolidate the two MockServer.  Thanks!

dtivel replied:
Yes, please create a tracking issue to consolidate.

heng-liu replied:
Thanks! Created tracking issue: https://github.com/NuGet/Client.Engineering/issues/949

heng-liu committed:
fix after rebasing

dominoFire reviewed (commented):
More comments on strings.

heng-liu committed:
address comments 5

dtivel commented on src/NuGet.Core/NuGet.CommandLine.XPlat/Commands/Signing/SignCommand.cs:
```diff
@@ -0,0 +1,233 @@
+            app.Command("sign", signCmd =>
+            {
+                CommandArgument packagePaths = signCmd.Argument(
+                    "<package-paths>",
+                    Strings.SignCommandPackagePathDescription,
+                    multipleValues: true);
```
Should we update the [spec](https://github.com/NuGet/Home/pull/9945/files#diff-62408b820dab5a1974ab95662420c9b2118b627ca47aa72769afc02e0fbb4324R50) (in multiple places) to reflect a possible plurality of values?

heng-liu replied:
Thanks! Added into https://github.com/NuGet/docs.microsoft.com-nuget/issues/2416 for tracking.

dtivel commented on src/NuGet.Core/NuGet.CommandLine.XPlat/Commands/Signing/SignCommand.cs:
```diff
@@ -0,0 +1,233 @@
+
+                signCmd.OnExecute(async () =>
+                {
+                    var logger = getLogger();
+
+                    ValidatePackagePaths(packagePaths);
+                    WarnIfNoTimestamper(logger, timestamper);
+                    ValidateCertificateInputs(path, fingerPrint, subject, store, location);
+                    ValidateAndCreateOutputDirectory(outputDirectory);
+
+                    var signingSpec = SigningSpecifications.V1;
+                    var storeLocation = ValidateAndParseStoreLocation(location);
+                    var storeName = ValidateAndParseStoreName(store);
+                    var hashAlgorithm = CommandLineUtility.ParseAndValidateHashAlgorithm(algorithm.Value(), algorithm.LongName, signingSpec);
+                    var timestampHashAlgorithm = CommandLineUtility.ParseAndValidateHashAlgorithm(timestamperAlgorithm.Value(), timestamperAlgorithm.LongName, signingSpec);
```
Use explicit types instead of `var`.

heng-liu replied:
Thanks! Fixed.

dtivel commented on src/NuGet.Core/NuGet.CommandLine.XPlat/Commands/Signing/SignCommand.cs:
```diff
@@ -0,0 +1,233 @@
+
+                    setLogLevel(XPlatUtility.MSBuildVerbosityToNuGetLogLevel(verbosity.Value()));
+
+                    var runner = getCommandRunner();
```
Use explicit type instead of `var`.

heng-liu replied:
Thanks! Fixed.

dtivel commented on src/NuGet.Core/NuGet.CommandLine.XPlat/Commands/Signing/SignCommand.cs:
```diff
@@ -0,0 +1,233 @@
+                    Strings.SignCommandCertificateSubjectNameDescription,
+                    CommandOptionType.SingleValue);
+
+                CommandOption fingerPrint = signCmd.Option(
```
`fingerprint` not `fingerPrint`.  Fingerprint is one word.

heng-liu replied:
Thanks! Fixed.

dtivel commented on src/NuGet.Core/NuGet.CommandLine.XPlat/Commands/Signing/SignCommand.cs:
```diff
@@ -0,0 +1,233 @@
+
+        private static StoreLocation ValidateAndParseStoreLocation(CommandOption location)
+        {
+            var storeLocation = StoreLocation.CurrentUser;
```
Use explicit type instead of `var`.

heng-liu replied:
Thanks! Fixed.

dtivel commented on src/NuGet.Core/NuGet.CommandLine.XPlat/Commands/Signing/SignCommand.cs:
```diff
@@ -0,0 +1,233 @@
+                    throw new ArgumentException(string.Format(CultureInfo.CurrentCulture,
+                        Strings.Err_InvalidValue,
+                        location.LongName,
+                        string.Join(",", Enum.GetValues(typeof(StoreLocation)).Cast<StoreLocation>().ToList())));
```
Simplify this:
```C#
Enum.GetValues(typeof(StoreLocation)).Cast<StoreLocation>().ToList()
```
...with this:
```C#
Enum.GetValues<StoreLocation>()
```

heng-liu replied:
Seems it doesn't work for me. I'm getting "[Compiler Error CS0308](https://docs.microsoft.com/en-us/dotnet/csharp/misc/cs0308)" if I simplify in this way..
`The non-generic method 'Enum.GetValues(Type)' cannot be used with type arguments`

dtivel commented on src/NuGet.Core/NuGet.CommandLine.XPlat/Commands/Signing/SignCommand.cs:
```diff
@@ -0,0 +1,233 @@
+                    throw new ArgumentException(string.Format(CultureInfo.CurrentCulture,
+                        Strings.Err_InvalidValue,
+                        store.LongName,
+                        string.Join(",", Enum.GetValues(typeof(StoreName)).Cast<StoreName>().ToList())));
```
Simplify this:
```C#
Enum.GetValues(typeof(StoreName)).Cast<StoreName>().ToList()
```
...with this:
```C#
Enum.GetValues<StoreName>()
```

heng-liu replied:
Seems it doesn't work for me. I'm getting "[Compiler Error CS0308](https://docs.microsoft.com/en-us/dotnet/csharp/misc/cs0308)" if I simplify in this way..
`The non-generic method 'Enum.GetValues(Type)' cannot be used with type arguments`

dtivel commented on src/NuGet.Core/NuGet.CommandLine.XPlat/Commands/Signing/SignCommand.cs:
```diff
@@ -0,0 +1,233 @@
+            return storeName;
+        }
+
+        private static void ValidateCertificateInputs(CommandOption path, CommandOption fingerPrint,
```
`fingerPrint` -> `fingerprint`

heng-liu replied:
Thanks! Fixed.

dtivel commented on src/NuGet.Core/NuGet.CommandLine.XPlat/Commands/Signing/SignCommand.cs:
```diff
@@ -0,0 +1,233 @@
+        internal static void Register(CommandLineApplication app,
+                         Func<ILogger> getLogger,
+                         Action<LogLevel> setLogLevel,
+                         Func<ISignCommandRunner> getCommandRunner)
```
Indent only 4 spaces from line 18.

heng-liu replied:
Thanks! Fixed.

dtivel commented on src/NuGet.Core/NuGet.CommandLine.XPlat/Commands/Signing/SignCommand.cs:
```diff
@@ -0,0 +1,233 @@
+
+        private static StoreName ValidateAndParseStoreName(CommandOption store)
+        {
+            var storeName = StoreName.My;
```
Use explicit type instead of `var`.

heng-liu replied:
Thanks! Fixed.

dtivel commented on src/NuGet.Core/NuGet.CommandLine.XPlat/Commands/Signing/SignCommand.cs:
```diff
@@ -0,0 +1,233 @@
+                         Action<LogLevel> setLogLevel,
+                         Func<ISignCommandRunner> getCommandRunner)
+        {
+            app.Command("sign", signCmd =>
```
Create a field:
```C#
private static readonly string CommandName = "sign";
```
...and use it here and [here](https://github.com/NuGet/NuGet.Client/pull/3974/files#diff-4cf20fb6a1c739714683d0b642b396837f22ccf7152372552d134b73cfddcad3R143).

heng-liu replied:
Thanks! Fixed. Changed into `private const string CommandName = "sign";`

dtivel commented on src/NuGet.Core/NuGet.CommandLine.XPlat/Utility/CommandLineUtility.cs:
```diff
@@ -36,5 +37,31 @@ public static string[] SplitAndJoinAcrossMultipleValues(IList<string> inputs)
+        /// <returns>Supported hash algorithm</returns>
+        internal static HashAlgorithmName ParseAndValidateHashAlgorithm(string optionValue, string optionName, SigningSpecifications spec)
+        {
+            var hashAlgorithm = HashAlgorithmName.SHA256;
```
Use explicit type instead of `var`.

heng-liu replied:
Thanks! Fixed.

dtivel commented on src/NuGet.Core/NuGet.CommandLine.XPlat/Utility/CommandLineUtility.cs:
```diff
@@ -36,5 +37,31 @@ public static string[] SplitAndJoinAcrossMultipleValues(IList<string> inputs)
+            {
+                throw new ArgumentException(string.Format(CultureInfo.CurrentCulture,
+                        Strings.Err_InvalidValue,
+                        optionName, string.Join(",", spec.AllowedHashAlgorithms)));
```
Indent 1 level (4 spaces) from line 59.

heng-liu replied:
Thanks! Fixed.

dtivel commented on src/NuGet.Core/NuGet.Commands/SignCommand/SignArgs.cs:
```diff
@@ -16,8 +18,33 @@ public class SignArgs
+        }
+
+        /// <summary>
+        /// Paths to the packages that has to be verified.
```
verified?  You mean signed.  "Paths to the packages that should be signed."

heng-liu replied:
Thanks! Fixed.

dtivel commented on src/NuGet.Core/NuGet.Commands/SignCommand/SignCommandRunner.cs:
```diff
@@ -22,38 +23,58 @@ public class SignCommandRunner : ISignCommandRunner
-            LocalFolderUtility.EnsurePackageFileExists(signArgs.PackagePath, packagesToSign);
-
-            var cert = await GetCertificateAsync(signArgs);
+            var packagesToSign = signArgs.PackagePaths.SelectMany(packagePath =>
```
Use explicit type instead of `var`.

heng-liu replied:
Thanks! Fixed.

dtivel commented on test/NuGet.Core.FuncTests/Dotnet.Integration.Test/DotnetSignTests.cs:
```diff
@@ -0,0 +1,523 @@
+        }
+
+        [Fact]
+        public async Task DotnetSign_SignPackageWithTrustedCertificate_SuccceedsAsync()
```
`Succeeds` not `Succceeds`.  Search and replace other occurrences.

heng-liu replied:
Thanks! Fixed.

dtivel commented on test/NuGet.Core.FuncTests/Dotnet.Integration.Test/DotnetSignTests.cs:
```diff
@@ -0,0 +1,523 @@
+                //Act
+                var result = _msbuildFixture.RunDotnet(
+                        pathContext.PackageSource,
+                        $"nuget sign .{Path.DirectorySeparatorChar}{packageFileName} --certificate-fingerprint {trustedCert.Source.Cert.Thumbprint} --certificate-store-name {trustedCert.StoreName} --certificate-store-location {trustedCert.StoreLocation}",
```
Indent 1 level (4 spaces) from line 76.  Check other places.

heng-liu replied:
Thanks! Fixed.

dtivel commented on test/NuGet.Core.FuncTests/Dotnet.Integration.Test/DotnetSignTests.cs:
```diff
@@ -0,0 +1,523 @@
+
+                TrustedTestCert<TestCertificate> trustedCert = _signFixture.TrustedTestCertificateChain.Leaf;
+                //Act
+                var result = _msbuildFixture.RunDotnet(
```
Use explicit type instead of `var`.  Check other places too.

heng-liu replied:
Thanks! Fixed.

dtivel commented on test/NuGet.Core.FuncTests/Dotnet.Integration.Test/MockServer.cs:
```diff
@@ -0,0 +1,440 @@
+    /// A Mock Server that is used to mimic a NuGet Server.
+    /// </summary>
+    [System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Design", "CA1001:TypesThatOwnDisposableFieldsShouldBeDisposable")]
+    public class MockServer : IDisposable
```
Can this be `internal` for now?

heng-liu replied:
I think so. Changed into `internal`. Thanks!

dtivel commented on test/NuGet.Core.FuncTests/Dotnet.Integration.Test/DotnetSignTests.cs:
```diff
@@ -0,0 +1,523 @@
+
+                TrustedTestCert<TestCertificate> trustedCert = _signFixture.TrustedTestCertificateChain.Leaf;
+                //Act
+                var result = _msbuildFixture.RunDotnet(
```
Use explicit type instead of `var`.  Also indent following lines 1 level (4 spaces).

Check whole file.

heng-liu replied:
Thanks! Fixed.

dtivel commented on test/NuGet.Core.Tests/NuGet.Commands.Test/CertificatesFixture.cs:
```diff
@@ -11,11 +11,20 @@ public sealed class CertificatesFixture : IDisposable
         {
             _defaultCertificate = SigningTestUtility.GenerateCertificate("test", generator => { });
+
+            var _defaultCertificateForTrust = SigningTestUtility.GenerateCertificate("test trusted", generator => { });
```
Use explicit type instead of `var`.

heng-liu replied:
Thanks! Fixed.

dtivel commented on test/NuGet.Core.Tests/NuGet.Commands.Test/CertificatesFixture.cs:
```diff
@@ -11,11 +11,20 @@ public sealed class CertificatesFixture : IDisposable
         {
             _defaultCertificate = SigningTestUtility.GenerateCertificate("test", generator => { });
+
+            var _defaultCertificateForTrust = SigningTestUtility.GenerateCertificate("test trusted", generator => { });
```
Not crazy about this name.  The name implies that `_defaultCertificate` is untrusted.  Is that true?  If yes, then rename to `_defaultUntrustedCertificate` and the associated method to `GetDefaultUntrustedCertificate()`.

Leading underscores are typically reserved for fields.  Remove it here.

`defaultTrustedCertificate` is more clear than `defaultCertificateForTrust`.  Rename associated method to `GetDefaultTrustedCertificate()`.

heng-liu replied:
Thanks! Fixed.

dtivel commented on test/TestUtilities/Test.Utility/Signing/CertificateStoreUtilities.cs:
```diff
@@ -18,5 +18,29 @@ public static StoreLocation GetTrustedCertificateStoreLocation(bool readOnly = f
+            return StoreName.My;
+        }
+
+        internal static StoreLocation GetTrustedCertificateStoreLocatioinForLeafOrSelfIssued()
```
`Locatioin` -> `Location`

Append `Certificate` to method name.

heng-liu replied:
Thanks! Fixed.

dtivel commented on test/TestUtilities/Test.Utility/Signing/CertificateStoreUtilities.cs:
```diff
@@ -18,5 +18,29 @@ public static StoreLocation GetTrustedCertificateStoreLocation(bool readOnly = f
+            return (RuntimeEnvironmentHelper.IsWindows || RuntimeEnvironmentHelper.IsLinux) ? StoreName.CertificateAuthority : StoreName.My;
+        }
+
+        internal static StoreName GetTrustedCertificateStoreNameForLeafOrSelfIssued()
```
Append `Certificate` to method name.

heng-liu replied:
Thanks! Added.

dtivel commented on test/TestUtilities/Test.Utility/TestServer/PortReserverOfMockServer.cs:
```diff
@@ -10,17 +12,17 @@
     /// </summary>
     [DebuggerDisplay("Port: {PortNumber}, Port count for this app domain: {_appDomainOwnedPorts.Count}")]
-    public class PortReserver : IDisposable
+    public class PortReserverOfMockServer : IDisposable
```
I don't understand this rename.  There's no `MockServer` dependency here.  Why not just keep the name `PortReserver`?

heng-liu replied:
Since there are already two different PortReserver.cs in dev branch before this PR:
https://github.com/NuGet/NuGet.Client/blob/dev/test/TestUtilities/Test.Utility/TestServer/PortReserver.cs
https://github.com/NuGet/NuGet.Client/blob/dev/test/NuGet.Clients.Tests/NuGet.CommandLine.Test/PortReserver.cs
The two PortReserver.cs have different APIs.
And Dotnet.Integration.Test in this PR needs a third one ( the same PortReserver with the second one).
So I move the second PortReserver into Test.Utility, but need to rename to avoid duplicate class name.

heng-liu replied:
I'll create another issue investigating if we could merge two PortResever, if it's necessary.
I agree with you that the naming is not good for now.

dominoFire reviewed (approved):
LGTM

heng-liu committed:
address comments 6

heng-liu commented:
Hi @kartheekp-ms , to address your offline question: if the functional tests "dotnet nuget sign" (in https://github.com/NuGet/NuGet.Client/blob/dev-hengliu-xplat3-feature-dotnet-nuget-sign/test/NuGet.Core.FuncTests/Dotnet.Integration.Test/DotnetSignTests.cs) is the same with "nuget sign" (in https://github.com/NuGet/NuGet.Client/blob/dev/test/NuGet.Clients.FuncTests/NuGet.CommandLine.FuncTest/Commands/SignCommandTests.cs).
I've compared the two, only the tests with interactive mode are not in "dotnet nuget sign".

heng-liu commented:
I've addressed almost all comments, and there is no blocking comments left. Pls let me know if it's good to sign-off this PR. @dtivel, @kartheekp-ms , @erdembayar . Thanks!
